File: app/src/main/java/app/kamy/qalbuApp/features/today/PrayerTrackerViewModel.kt
package app.kamy.qalbuApp.features.today

import android.content.Context
import androidx.lifecycle.ViewModel
import app.kamy.qalbuApp.domain.model.OptionalWorshipHabit
import app.kamy.qalbuApp.domain.model.PrayerType
import app.kamy.qalbuApp.domain.prayer.HijriDayHelper
import app.kamy.qalbuApp.domain.prayer.PrayerTrackerAvailability
import app.kamy.qalbuApp.infrastructure.notifications.PrayerCheckReminderScheduler
import app.kamy.qalbuApp.infrastructure.preferences.PrayerDayProgress
import app.kamy.qalbuApp.infrastructure.preferences.PrayerTrackerPreferencesStore
import app.kamy.qalbuApp.infrastructure.preferences.PrayerTrackerStore
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import java.util.Calendar
import javax.inject.Inject

data class OptionalHabitUiItem(
    val habit: OptionalWorshipHabit,
    val labelRes: Int,
    val completed: Boolean,
    val applicableToday: Boolean
)

data class PrayerTrackerUiState(
    val todayProgress: PrayerDayProgress = PrayerDayProgress(PrayerTrackerStore.todayKey(), 0),
    val weekProgress: List<PrayerDayProgress> = emptyList(),
    val monthPreview: List<PrayerDayProgress> = emptyList(),
    val streak: Int = 0,
    val bestStreak: Int = 0,
    val challengeTarget: Int = 7,
    val completedPrayers: Set<PrayerType> = emptySet(),
    val optionalHabits: List<OptionalHabitUiItem> = emptyList(),
    val availablePrayers: Set<PrayerType> = emptySet(),
    val checkRemindersEnabled: Boolean = true
)

@HiltViewModel
class PrayerTrackerViewModel @Inject constructor(
    @ApplicationContext private val appContext: Context
) : ViewModel() {

    private val prefs = PrayerTrackerPreferencesStore.from(appContext)

    private val _state = MutableStateFlow(PrayerTrackerUiState())
    val state: StateFlow<PrayerTrackerUiState> = _state.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        val today = PrayerTrackerStore.todayKey()
        val completed = PrayerTrackerStore.TRACKED_PRAYERS
            .filter { PrayerTrackerStore.isCompleted(appContext, it, today) }
            .toSet()
        val streak = PrayerTrackerStore.currentStreak(appContext)
        val cal = Calendar.getInstance()
        val available = PrayerTrackerAvailability.availablePrayers(appContext)
        _state.update {
            it.copy(
                todayProgress = PrayerTrackerStore.dayProgress(appContext, today),
                weekProgress = PrayerTrackerStore.weekProgress(appContext),
                monthPreview = PrayerTrackerStore.monthProgress(
                    appContext,
                    cal.get(Calendar.YEAR),
                    cal.get(Calendar.MONTH) + 1
                ),
                streak = streak,
                bestStreak = PrayerTrackerStore.bestStreak(appContext),
                challengeTarget = PrayerTrackerStore.challengeTargetDays(streak),
                completedPrayers = completed,
                optionalHabits = buildOptionalHabits(today),
                availablePrayers = available,
                checkRemindersEnabled = prefs.checkRemindersEnabled()
            )
        }
    }

    fun togglePrayer(prayer: PrayerType) {
        val completed = _state.value.completedPrayers.contains(prayer)
        if (!PrayerTrackerAvailability.canToggle(appContext, prayer, completed)) return
        PrayerTrackerStore.toggle(appContext, prayer)
        if (PrayerTrackerStore.isCompleted(appContext, prayer)) {
            PrayerCheckReminderScheduler.onPrayerMarked(appContext, prayer)
        } else {
            PrayerCheckReminderScheduler.reschedule(appContext)
        }
        refresh()
    }

    fun toggleOptionalHabit(habit: OptionalWorshipHabit) {
        PrayerTrackerStore.toggleOptional(appContext, habit)
        refresh()
    }

    fun setCheckRemindersEnabled(enabled: Boolean) {
        prefs.setCheckRemindersEnabled(enabled)
        if (enabled) {
            PrayerCheckReminderScheduler.reschedule(appContext)
        } else {
            PrayerCheckReminderScheduler.cancelAll(appContext)
        }
        refresh()
    }

    fun isCompleted(prayer: PrayerType): Boolean =
        _state.value.completedPrayers.contains(prayer)

    private fun buildOptionalHabits(today: String): List<OptionalHabitUiItem> {
        val hijriDay = HijriDayHelper.todayHijriDay(appContext)
            ?: HijriDayHelper.parseHijriDayFromLabel(
                app.kamy.qalbuApp.infrastructure.cache.PrayerDayCache.load(appContext)?.hijriLabel
            )
        val monThu = HijriDayHelper.isMondayThursdayFastDay()
        val ayyamulBidh = HijriDayHelper.isAyyamulBidh(hijriDay)

        return OptionalWorshipHabit.ALL.mapNotNull { habit ->
            if (!prefs.isOptionalHabitEnabled(habit)) return@mapNotNull null
            val applicable = when (habit) {
                OptionalWorshipHabit.QIYAMUL_LAIL -> true
                OptionalWorshipHabit.MONDAY_THURSDAY_FAST -> monThu
                OptionalWorshipHabit.AYYAMUL_BIDH_SAHUR -> ayyamulBidh
            }
            if (!applicable) return@mapNotNull null
            OptionalHabitUiItem(
                habit = habit,
                labelRes = habit.labelRes(),
                completed = PrayerTrackerStore.isOptionalCompleted(appContext, habit, today),
                applicableToday = applicable
            )
        }
    }

    private fun OptionalWorshipHabit.labelRes(): Int = when (this) {
        OptionalWorshipHabit.QIYAMUL_LAIL -> app.kamy.qalbuApp.R.string.optional_qiyam
        OptionalWorshipHabit.MONDAY_THURSDAY_FAST -> app.kamy.qalbuApp.R.string.optional_mon_thu_fast
        OptionalWorshipHabit.AYYAMUL_BIDH_SAHUR -> app.kamy.qalbuApp.R.string.optional_ayyamul_bidh
    }
}

File: iosApp/Saat/Features/Discovery/ViewModels/PrayerTrackerViewModel.swift
//
//  PrayerTrackerViewModel.swift
//  Saat
//
//  Created by Elmee on 25/06/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Combine
import Foundation

internal struct OptionalHabitUiItem: Identifiable, Sendable {
    internal var id: String { habit.rawValue }
    internal let habit: OptionalWorshipHabit
    internal let label: String
    internal let completed: Bool
    internal let applicableToday: Bool
}

internal struct PrayerTrackerUiState: Sendable {
    internal var todayProgress: PrayerDayProgress = PrayerDayProgress(dayKey: "", completedCount: 0, totalCount: 5, optionalCompletedCount: 0, optionalTotalCount: 0)
    internal var weekProgress: [PrayerDayProgress] = []
    internal var monthPreview: [PrayerDayProgress] = []
    internal var streak: Int = 0
    internal var bestStreak: Int = 0
    internal var challengeTarget: Int = 7
    internal var completedPrayers: Set<PrayerType> = []
    internal var optionalHabits: [OptionalHabitUiItem] = []
    internal var availablePrayers: Set<PrayerType> = []
}

@MainActor
internal final class PrayerTrackerViewModel: ObservableObject {
    @Published internal var state = PrayerTrackerUiState()
    @Published internal var toastMessage: String?

    private let store: PrayerTrackerStore
    private let controller: PrayerTimesController
    private var controllerCancellable: AnyCancellable?
    private var languageCancellable: AnyCancellable?

    internal init(appGroupIdentifier: String?, controller: PrayerTimesController) {
        self.store = PrayerTrackerStore(appGroupIdentifier: appGroupIdentifier)
        self.controller = controller

        self.controllerCancellable = controller.objectWillChange
            .receive(on: RunLoop.main)
            .sink { [weak self] _ in
                self?.refresh()
            }
        
        self.languageCancellable = NotificationCenter.default.publisher(for: .appLanguageDidChange)
            .receive(on: RunLoop.main)
            .sink { [weak self] _ in
                self?.refresh()
            }
        
        refresh()
    }

    internal func refresh() {
        let today = store.todayKey()
        let completed = PrayerTrackerStore.TRACKED_PRAYERS.filter {
            store.isCompleted(prayer: $0, dayKey: today)
        }
        let streak = store.currentStreak()

        let cal = Calendar.current
        let year = cal.component(.year, from: Date())
        let month = cal.component(.month, from: Date())

        let available = availablePrayersList()

        self.state = PrayerTrackerUiState(
            todayProgress: store.dayProgress(dayKey: today),
            weekProgress: store.weekProgress(),
            monthPreview: store.monthProgress(year: year, month: month),
            streak: streak,
            bestStreak: store.bestStreak(),
            challengeTarget: store.challengeTargetDays(streak: streak),
            completedPrayers: Set(completed),
            optionalHabits: buildOptionalHabits(today: today),
            availablePrayers: Set(available)
        )
    }

    internal func togglePrayer(_ prayer: PrayerType) {
        let completed = state.completedPrayers.contains(prayer)
        // Check availability
        if !completed && !availablePrayersList().contains(prayer) {
            // Cannot mark future prayer
            return
        }

        let today = store.todayKey()
        _ = store.toggle(prayer: prayer, dayKey: today)
        
        let nowCompleted = store.isCompleted(prayer: prayer, dayKey: today)
        if nowCompleted {
            let prayerName = AppLanguageManager.shared.localize("prayer_" + prayer.rawValue.lowercased())
            self.toastMessage = String(format: AppLanguageManager.shared.localize("toast_marked_completed"), prayerName)
        }
        refresh()
    }

    internal func toggleOptionalHabit(_ habit: OptionalWorshipHabit) {
        let today = store.todayKey()
        _ = store.toggleOptional(habit: habit, dayKey: today)
        refresh()
    }

    private func availablePrayersList() -> [PrayerType] {
        let now = Date()
        // Map dailyPrayers from PrayerTimesController
        var list: [PrayerType] = []
        for entry in controller.dailyPrayers {
            if let type = PrayerType.from(aladhanKey: entry.name) {
                if now >= entry.date {
                    list.append(type)
                }
            }
        }
        // If dailyPrayers is empty, allow checking based on time or fallback to all
        if controller.dailyPrayers.isEmpty {
            return [.fajr, .dhuhr, .asr, .maghrib, .isha]
        }
        return list
    }

    private func buildOptionalHabits(today: String) -> [OptionalHabitUiItem] {
        // In iOS, we enable Qiyamul Lail by default, and others are optional
        // For Monday/Thursday fasting: check day of week (2 is Monday, 5 is Thursday in Calendar)
        let weekday = Calendar.current.component(.weekday, from: Date())
        let isMonThu = (weekday == 2 || weekday == 5)

        // For Ayyamul Bidh (13, 14, 15 of Hijri month).
        // Since we have hijriDateLabel (e.g. "13 Shaban 1445"), let's parse the day number
        var isAyyamulBidh = false
        if let hijri = controller.hijriDateLabel {
            let num = hijri.split(separator: " ").first.flatMap { Int($0) }
            if let dayVal = num, (13...15).contains(dayVal) {
                isAyyamulBidh = true
            }
        }

        return OptionalWorshipHabit.allCases.compactMap { habit in
            let applicable = (habit == .qiyamulLail) ? true :
                             (habit == .mondayThursdayFast ? isMonThu : isAyyamulBidh)
            
            guard applicable else { return nil }

            let label = (habit == .qiyamulLail) ? AppLanguageManager.shared.localize("habit_qiyamul_lail") :
                        (habit == .mondayThursdayFast ? AppLanguageManager.shared.localize("habit_monday_thursday_fast") : AppLanguageManager.shared.localize("habit_ayyamul_bidh_fast"))

            return OptionalHabitUiItem(
                habit: habit,
                label: label,
                completed: store.isOptionalCompleted(habit: habit, dayKey: today),
                applicableToday: true
            )
        }
    }
}
